### PR TITLE
Fix Exercise 5-01

### PR DESCRIPTION
**Pull request by ibnIrshad**
Daniel,

Another spec nitpick! Hope it makes sense. 

I'm still learning this forking business, I tried to create a new commit on my fork (after it diverged), and then pull it back to your repo without a pesky merge commit, but couldn't figure it out. Gave up eventually and made my one line change using Github's web interface.

Anyhow here it is.

### CODE REVIEW

**Comment by sporkbox** [comment]
If you check `master`, you'll see that when I accepted your pull request it resulted in a merge commit. That's a normal occurrence. That's because Pull Requests are basically asking the original author to merge in changes from another branch.

I'm not sure how familiar you are with Git, but each commit generally only has one parent, which creates a linear history. A commit that has two parent IDs is a merge commit, since logistically it really *did* come from the original branch and the contributing branch. Thus, there really isn't a good way to merge two branches without creating that merge commit.

Since we're collaborating over GitHub, keeping changes on a separate branch and doing pull requests is likely the smoother way to do things here. With that said, let's talk code. :)

---

Upon looking at the 5-01 solution again, I'm wondering if there isn't a better way for us to accomplish this. The instructions ask us to fix `getint` by pushing the non-digit input back onto the buffer if it occurs after a `+` or `-`, but even with your correction we're still returning zero on any non-numeric input, independent of the presence of `+` or `-`. Thankfully, I'm pretty sure we can meet spec better by moving the `ungetch` and `return` to the if statement like so:

    if (c == '+' || c == '-') {
        c = getch();
        if (!isdigit(c)) {
            ungetch(sign == 1 ? '+' : '-');
            return 0;
        }
    }

You can add another commit to your branch (or again, try using `--amend`) and it 
should show up here in the pull request, ready for me to merge it. Or I can 
apply the edit myself in `master` and I believe we're able to close or cancel a 
pull request. It depends on which approach you'd prefer. I'm okay with either. 
:)

**Comment by ibnIrshad** [comment]
Ah I see what you're trying to do, putting the sign character itself back on the buffer. I was trying to put only the non-digit character _after_ the sign char back on the buffer. However I think your correction can be improved once more: we need to add another `ungetch(c)` line before we `ungetch` the sign character, like so:

```
if (c == '+' || c == '-') {
    c = getch();
    if (!isdigit(c)) {
        ungetch(c); //we need to do this too
        ungetch(sign == 1 ? '+' : '-');
        return 0;
    }
}
```


On page 96 of K&R (the page before this exercise) the spec says
> Our version of `getint` returns `EOF` for end of file, zero if the next input is not a number, and a positive value if the input contains a valid number.

So there are two requirements for this program:
- return a value as specified in the quote above
- preserve the buffer contents in case of invalid input

Just in order to clarify things, and be able to test the program's correctness, I'm also going to print the contents of the buffer in `main` after the program exits.

I've modified the PR, what do you think?

**Comment by sporkbox** [comment]
I'm not sure what the effect would be, in practice, to `ungetch` the non-numeric character, but it's still within spec and may actually be closer to what was intended!

I think it's fine except for one little issue: `buf` just stores characters.
That means if we're going to `printf` its contents, we need to make sure it's
NUL terminated or we'll run into undefined behavior. I decided to look into the
reference at the end of the book to make sure, and I found it on page 219 in the
2nd Edition:

> A static object not explicitly initialized is initialized as if it (or its 
> members) were assigned the constant `0`. The initial value of an automatic 
> object not explicitly initialized is undefined.

I checked a little further to look into `buf`'s specific situation: it's a 
global and is available to all functions, so I wanted to know if globals are 
considered static or not. Page 195 had the answer, inside §A4.1:

> The objects declared outside all blocks, at the same level as function 
> definitions, are always static.

So that means in this case I'm technically wrong, because `buf` was already 
initialized with NUL bytes. Even still, I decided I'd write this up anyway since 
I learned something and maybe it'll help you, too.

---

The TLDR is "globals are considered static, and static vars will always be 
initialized to zero." So coming full circle, the PR is good and I'll go ahead 
and merge. One last thing, though: do you want to add a line in AUTHORS 
crediting you for this? If you want to do that, I'll wait to merge since I'd 
rather have credits like that accompany their commit so history looks clean.

Thanks again for contributing!

**Comment by ibnIrshad** [comment]
Interesting stuff, good to know.

I'm not sure whether the convention is to keep the AUTHORS edits in the very same commit or not, but I did.

**Comment by sporkbox** [comment]
Based on what I've learned from other projects, if an author adds something to the project, their edits are noted in the same commit that has their proposed changes.

I just wanted to point out that it's okay to have multiple lines under your name in AUTHORS. I don't know if you felt you needed to condense it, or if it was just a tone change you preferred, but I wanted to make that clear. :smile:

Here's hoping I can get the spare time to clean up the project's code and commenting style over the weekend!
